io/eolearn/tests/test_io.py

```diff
@@ -321,7 +345,192 @@ def test_feature_dimension(self):
     def test_time_interval_datetime(self):
         for task in self.task_cases_datetime:
             with self.subTest(msg='Test case {}'.format(task.name)):
-                self.assertEqual(task.eop.meta_info['time_interval'], self.time_interval_datetime)
+                self.assertEqual(task.eop.meta_info['time_interval'], task.time_interval)
+
+    def test_bbox(self):
+        for task in self.task_cases:
+            with self.subTest(msg='Test case {}'.format(task.name)):
+                self.assertEqual(task.eop.bbox, task.bbox)
+
+
+class TestTimelessFeatures(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.res_x = '53m'
+        cls.res_y = '78m'
+        cls.instance_id = os.environ.get('INSTANCE_ID')
+        cls.time_interval = ('2018-1-5', '2018-1-6')
+        cls.bbox = BBox(bbox=(-5.05, 48.0, -5.00, 48.05), crs=CRS.WGS84)
+        cls.dem_request = DEMWCSInput(
+            layer='DEM',
+            resx=cls.res_x,
+            resy=cls.res_y,
+            instance_id=cls.instance_id
+        )
+        cls.timed_request = SentinelHubWCSInput(
+            layer='BANDS-S2-L1C',
+            resx=cls.res_x,
+            resy=cls.res_y,
+            data_source=DataSource.SENTINEL2_L1C,
+            instance_id=cls.instance_id
+        )
+
+    def test_dem_to_empty_eopatch(self):
+        with self.subTest(msg='Test to add DEM (timeless) to empty eopatch'):
+            data = self.dem_request.execute(bbox=self.bbox)
+            self.assertTrue(len(data.timestamp) == 0)
+            self.assertFalse(data.meta_info['time_interval'])
+
+    def test_dem_to_eopatch_with_timestamps(self):
+        with self.subTest(msg='Test to add DEM (timeless) to eopatch with non-empty timestamps'):
+            timed_eopatch = self.timed_request.execute(bbox=self.bbox, time_interval=self.time_interval)
+            prior_timestamps = timed_eopatch.timestamp.copy()
+            dem_eopatch = self.dem_request.execute(eopatch=timed_eopatch, bbox=self.bbox)
+            self.assertListEqual(dem_eopatch.timestamp, prior_timestamps,
+                                 'Timestamps before and after adding dem differ!')
+            self.assertTrue(dem_eopatch.meta_info['time_interval'], '\'time_interval\' in meta_info is expected!')
+
+    def test_time_data_to_dem_eopatch(self):
+        with self.subTest(msg='Test to add timed data to DEM (timeless) eopatch'):
+            dem_eopatch = self.dem_request.execute(bbox=self.bbox)
+            timed_eopatch = self.timed_request.execute(eopatch=dem_eopatch, bbox=self.bbox,
+                                                       time_interval=self.time_interval)
+            self.assertTrue(len(timed_eopatch.timestamp) == 1, 'timed_eopatch should have 1 timestamp!')
+            self.assertTrue(timed_eopatch.meta_info['time_interval'], '\'time_interval\' in meta_info is expected!')
+
+
+class TestImageFormats(unittest.TestCase):

```

Comment:
I haven't checked all the details but it looks to me there is some code repetition between `TestEOPatch` and `TestImageFormats` classes.
Couldn't you just add `test_image_format` function to `TestEOPatch` class? (That way you also reduce the amount of data needed to download.)

Btw, feel free to rename `TestEOPatch` class, not very good name for here.

Reply:
I had two purposes when working on tests:
* reduce downloaded data from SH (and making the tests faster). In order to do that I've picked a smaller BBox and defined much narrower `time_interval` (even split tests into two distinct sets, one with single timestamp and one with a few)
* make sure that refactoring `image_format` will result with equal results.
For the second task I've added `TestImageFormats`, where for each test two EOpatches are created and then compared. This is different from `TestEOPatch` where the comparison is done more-or-less on metadata of the resulting EOPatch.

Reply:
Additionally, `TestTimelessFeatures` test was added to test the fix in the code when temporal data is being added to completely timeless EOPatch (which was not working before, but is allowed now).

Reply:
Ok, I agree with this.